pkg/minibroker/provider.go

```diff
@@ -17,53 +17,115 @@ limitations under the License.
 type Provider interface {
-	Bind(service []corev1.Service, params map[string]interface{}, chartSecrets map[string]interface{}) (*Credentials, error)
+	Bind(
+		service []corev1.Service,
+		bindParams BindParams,
+		provisionParams ProvisionParams,
+		chartSecrets Object,
+	) (Object, error)
 }
 
-type Credentials struct {
-	Protocol string
-	URI      string `json:"uri,omitempty"`
-	Username string `json:"username,omitempty"`
-	Password string `json:"password,omitempty"`
-	Host     string `json:"host,omitempty"`
-	Port     int32  `json:"port,omitempty"`
-	Database string `json:"database,omitempty"`
+// Object is a wrapper around map[string]interface{} that implements methods for helping with
+// digging and type asserting.
+type Object map[string]interface{}
+
+var (
+	// ErrDigNotFound is the error for a key not found in the Object.
+	ErrDigNotFound = fmt.Errorf("key not found")
+	// ErrDigNotString is the error for a key that is not a string.
+	ErrDigNotString = fmt.Errorf("key is not a string")
+)
+
+// Dig digs the Object based on the provided key.
+// key must be in the format "foo.bar.baz". Each segment represents a level in the Object.
+func (o Object) Dig(key string) (interface{}, bool) {
+	keyParts := strings.Split(key, ".")
+	var part interface{} = o
+	var ok bool
+	for _, keyPart := range keyParts {
+		switch p := part.(type) {
+		case map[string]interface{}:
+			part, ok = p[keyPart]
+			if !ok {
+				return nil, false
+			}
+		case Object:
+			part, ok = p[keyPart]
+			if !ok {
+				return nil, false
+			}
+		default:
+			return nil, false
+		}
+	}
+	return part, ok
 }
 
-// ToMap converts the credentials into the OSB API credentials response
-// see https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#device-object
-// {
-//   "credentials": {
-//     "uri": "mysql://mysqluser:pass@mysqlhost:3306/dbname",
-//     "username": "mysqluser",
-//     "password": "pass",
-//     "host": "mysqlhost",
-//     "port": 3306,
-//     "database": "dbname"
-//     }
-// }
-func (c Credentials) ToMap() map[string]interface{} {
-	var result map[string]interface{}
-	j, _ := json.Marshal(c)
-	json.Unmarshal(j, &result)
-	return result
+// DigString wraps Object.Dig and type-asserts the found key.
+func (o Object) DigString(key string) (string, error) {
+	val, ok := o.Dig(key)
+	if !ok {
+		return "", ErrDigNotFound
+	}
+	valStr, ok := val.(string)
+	if !ok {
+		return "", ErrDigNotString
+	}
+	return valStr, nil
 }
 
-func buildURI(c Credentials) string {
-	if c.Database == "" {
-		return fmt.Sprintf("%s://%s:%s@%s:%d",
-			c.Protocol, c.Username, c.Password, c.Host, c.Port)
+// DigStringOr wraps Object.DigString and returns defaultValue if any error is returned from
+// Object.DigString.
+func (o Object) DigStringOr(key string, defaultValue string) string {
+	str, err := o.DigString(key)
+	if err != nil {
+		return defaultValue
 	}
+	return str
+}
+
+// BindParams is a new type that wraps Object.
+type BindParams Object
```

Comment:
If we do:
```suggestion
type BindParams struct {
	Object
}
```
(That is, if we embed `Object` instead), then we don't need to duplicate all the methods.

Reply:
The usage becomes less fluent but I'll make the switch.

Reply:
Fixed with 1cb2a49.